src/tracing/instrumentation/mongodb.js

```diff
@@ -55,54 +62,74 @@ function instrument(mongodb) {
     return;
   }
 
-  var traceId = event.operationId.traceId;
-  var parentSpanId = event.operationId.parentSpanId;
   var parentSpan = cls.getCurrentSpan();
+  if (parentSpan && cls.isExitSpan(parentSpan)) {
+    return;
+  }
+
+  var traceId;
+  var parentSpanId;
+  // event.operationId is only present prior to mongodb package version 3.0.6, see above.
+  if (event.operationId) {
+    traceId = event.operationId.traceId;
+    parentSpanId = event.operationId.parentSpanId;
+    supportsOperationIds = true;
+  }
+
   if (parentSpan && (traceId == null || parentSpanId == null)) {
+    // either event.operationId has not been present or event.operationId did not have a traceId/parentSpanId set.
     traceId = parentSpan.t;
     parentSpanId = parentSpan.s;
   }
 
   if (traceId == null || parentSpanId == null) {
-    // no parent found. We don't want to keep track of isolated mongodb spans
+    // We could not find the trace ID/parent span ID, neither via the operation ID mechanism nor the getCurrentSpan()
+    // way - give up finally.
     return;
   }
 
   var span = cls.startSpan('mongo', traceId, parentSpanId, false);
 
-  var host = event.connectionId.host;
-  var port = event.connectionId.port;
+  var peer = null;
+  var service = null;
+  if (event.connectionId && (event.connectionId.host || event.connectionId.port)) {
+    peer = {
+      hostname: event.connectionId.host,
+      port: event.connectionId.port,
+    };
+    service = event.connectionId.host + ':' + event.connectionId.port;
+  } else if (typeof event.connectionId === 'string') {
+    peer = parseConnectionToPeer(event.connectionId);
+    service = event.connectionId;
+  }
   var database = event.databaseName;
   var collection = event.command.collection || event.command[event.commandName];
-  // using the Mongodb instrumentation API, it is not possible to gather stack traces. Getting started
-  // without stack traces at first.
+  // using the Mongodb instrumentation API, it is not possible to gather stack traces.
   span.stack = [];
   span.data = {
-      peer: {
-        hostname: host,
-        port: port
-      },
-      mongo: {
-        command: event.commandName,
-        service: host + ':' + port,
-        namespace: database + '.' + collection,
-        filter: stringifyWhenNecessary(event.command.filter),
-        query: stringifyWhenNecessary(event.command.query)
-      }
-    };
-
-  event.operationId.traceId = span.t;
-  event.operationId.parentSpanId = span.p;
-
-  requests[getUniqueRequestId(event)] = {
-    span: span
+    peer: peer,
+    mongo: {
+      command: event.commandName,
+      service: service,
+      namespace: database + '.' + collection,
+      filter: stringifyWhenNecessary(event.command.filter),
+      query: stringifyWhenNecessary(event.command.query)
+    }
   };
+
+  if (event.operationId) {
+    event.operationId.traceId = span.t;
+    event.operationId.parentSpanId = span.p;
+  }
+  requests[getUniqueRequestId(event)] = span;
```

Comment:
Before this PR, we always unnecessarily wrapped the span in another object, this wrapping has been removed.